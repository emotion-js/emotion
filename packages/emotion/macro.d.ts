declare module 'emotion/macro' {
  export * from 'emotion'
}
