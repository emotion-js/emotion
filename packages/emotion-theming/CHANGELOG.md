# emotion-theming

## 10.0.27

### Patch Changes

- [`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968) [#1698](https://github.com/emotion-js/emotion/pull/1698) Thanks [@Andarist](https://github.com/Andarist)! - Add LICENSE file
- Updated dependencies [[`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968)]:
  - @emotion/core@10.0.27
  - @emotion/styled@10.0.27
  - @emotion/weak-memoize@0.2.5

## 10.0.19

### Patch Changes

- [ffc7c58c](https://github.com/emotion-js/emotion/commit/ffc7c58c13277c8f05861e2534fc2a85fff01b76) [#1509](https://github.com/emotion-js/emotion/pull/1509) Thanks [@XeeD](https://github.com/XeeD)! - Add TypeScript type definition for the useTheme hook in emotion-theming

- Updated dependencies [c81c0033]:
  - @emotion/weak-memoize@0.2.4

## 10.0.18

### Patch Changes

- [bfb40050](https://github.com/emotion-js/emotion/commit/bfb400503d0d8d399e4e0051dc6e5eac40624b10) [#1499](https://github.com/emotion-js/emotion/pull/1499) Thanks [@tkh44](https://github.com/tkh44)! - Add useTheme React hook to emotion-theming

## 10.0.17

### Patch Changes

- [66cda641](https://github.com/emotion-js/emotion/commit/66cda64128631790b81e3c9df273a972358ea593) [#1478](https://github.com/emotion-js/emotion/pull/1478) Thanks [@Andarist](https://github.com/Andarist)! - Update Babel dependencies

## 10.0.14

### Patch Changes

- [c0eb604d](https://github.com/emotion-js/emotion/commit/c0eb604d) [#1419](https://github.com/emotion-js/emotion/pull/1419) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Update build tool
  - [c673e200](https://github.com/emotion-js/emotion/commit/c673e200) [#1405](https://github.com/emotion-js/emotion/pull/1405) Thanks [@ryanswanson](https://github.com/ryanswanson)! - PropsOf<C> TypeScript utilities - Support defaultProps by delegating to new JSX and React types.
  - [33803731](https://github.com/emotion-js/emotion/commit/33803731) [#1397](https://github.com/emotion-js/emotion/pull/1397) Thanks [@SimeonC](https://github.com/SimeonC)! - Fixing flow types for the withTheme HOC
