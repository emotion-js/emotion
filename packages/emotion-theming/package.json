{
  "name": "emotion-theming",
  "version": "10.0.19",
  "description": "A CSS-in-JS theming solution, inspired by styled-components",
  "main": "dist/emotion-theming.cjs.js",
  "module": "dist/emotion-theming.esm.js",
  "types": "types/index.d.ts",
  "files": [
    "src",
    "dist",
    "types"
  ],
  "scripts": {
    "test:typescript": "dtslint types"
  },
  "repository": "https://github.com/emotion-js/emotion/tree/master/packages/emotion-theming",
  "keywords": [
    "react",
    "theme",
    "theming",
    "emotion",
    "cssinjs",
    "css-in-js"
  ],
  "author": "styled-components team",
  "contributors": [
    "Evan Scott <probablyup@gmail.com> (https://github.com/probablyup)"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/emotion-js/emotion/issues"
  },
  "homepage": "https://emotion.sh",
  "devDependencies": {
    "@emotion/core": "^10.0.17",
    "@emotion/styled": "^10.0.17",
    "@types/react": "^16.8.20",
    "dtslint": "^0.3.0",
    "react": "^16.5.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.5.5",
    "@emotion/weak-memoize": "0.2.4",
    "hoist-non-react-statics": "^3.3.0"
  },
  "peerDependencies": {
    "@emotion/core": "^10.0.17",
    "react": ">=16.3.0"
  },
  "umd:main": "dist/emotion-theming.umd.min.js",
  "browser": {
    "./dist/emotion-theming.cjs.js": "./dist/emotion-theming.browser.cjs.js",
    "./dist/emotion-theming.esm.js": "./dist/emotion-theming.browser.esm.js"
  },
  "preconstruct": {
    "umdName": "emotionTheming"
  }
}