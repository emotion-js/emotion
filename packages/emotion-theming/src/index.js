// @flow
export { default as ThemeProvider } from './theme-provider'
export { default as withTheme } from './with-theme'
export { channel, contextTypes, createBroadcast } from './utils'
