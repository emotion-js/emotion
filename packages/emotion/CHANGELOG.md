# emotion

## 10.0.27

### Patch Changes

- [`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968) [#1698](https://github.com/emotion-js/emotion/pull/1698) Thanks [@Andarist](https://github.com/Andarist)! - Add LICENSE file
- Updated dependencies [[`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968)]:
  - babel-plugin-emotion@10.0.27
  - create-emotion@10.0.27
  - @emotion/stylis@0.8.5

## 10.0.23

### Patch Changes

- [`1ae94891`](https://github.com/emotion-js/emotion/commit/1ae948917326e2bc2bc978c59d04cb2695c72e38) [#1583](https://github.com/emotion-js/emotion/pull/1583) Thanks [@Andarist](https://github.com/Andarist)! - Export TS & Flow types for emotion/macro

- Updated dependencies [[`3927293d`](https://github.com/emotion-js/emotion/commit/3927293d0b9d96b4a7c00196e8430728759b1161), [`b3a0f148`](https://github.com/emotion-js/emotion/commit/b3a0f1484f2efcc78b447639ff2e0bc0f29915ae)]:
  - babel-plugin-emotion@10.0.23

## 10.0.17

### Patch Changes

- [66cda641](https://github.com/emotion-js/emotion/commit/66cda64128631790b81e3c9df273a972358ea593) [#1478](https://github.com/emotion-js/emotion/pull/1478) Thanks [@Andarist](https://github.com/Andarist)! - Add warnings about using illegal escape sequences

## 10.0.14

### Patch Changes

- [c0eb604d](https://github.com/emotion-js/emotion/commit/c0eb604d) [#1419](https://github.com/emotion-js/emotion/pull/1419) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Update build tool
