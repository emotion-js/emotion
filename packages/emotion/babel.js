module.exports = require('babel-plugin-emotion')
