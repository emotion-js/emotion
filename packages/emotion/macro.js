module.exports = require('babel-plugin-emotion').macros.emotion
