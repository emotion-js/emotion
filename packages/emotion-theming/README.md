# emotion-theming

> A CSS-in-JS theming solution for React

_`emotion-theming` is a theming library inspired by [styled-components](https://github.com/styled-components/styled-components)_

## Table of Contents

- [Install](#install)
- [Usage](#usage)
- [API](#api)
  - [ThemeProvider](#themeprovider-reactcomponenttype)
  - [withTheme](#withthemecomponent-reactcomponenttype-reactcomponenttype)
- [Credits](#credits)
- [License](#license)

## Install

```bash
# add --save if using npm@4 or lower
npm i emotion-theming

# or
yarn add emotion-theming
```

## Usage

Theming is accomplished by placing the `ThemeProvider` component, at the top of the React component tree and wrapping descendants with the `withTheme` higher-order component. This HOC gets the current theme and injects it as a "prop" into your own component.

The theme prop is automatically injected into components created with `styled`. The theme can also be accessed via passing a function to the css prop.

```jsx
// Page.js
import * as React from 'react'
/** @jsx jsx */
import { jsx } from '@emotion/core'
import styled from '@emotion/styled'

const Container = styled.div({
  background: 'whitesmoke',
  height: '100vh'
})

const Headline = styled.h1`
  color: ${props => props.theme.color};
  font-family: sans-serif;
`

export default class Page extends React.Component {
  render() {
    return (
      <Container>
        <Headline>I'm red!</Headline>
        <p css={theme => ({ color: theme.color })}>I'm also red!</p>
      </Container>
    )
  }
}

// App.js
import React from 'react'
import ReactDOM from 'react-dom'
import { ThemeProvider } from 'emotion-theming'

import Page from './Page.js'

const theme = {
  color: 'red'
}

class App extends React.Component {
  render() {
    return (
      <ThemeProvider theme={theme}>
        <Page />
      </ThemeProvider>
    )
  }
}
```

## API

### ThemeProvider: React.ComponentType

A React component that passes the theme object down the component tree via [context](https://reactjs.org/docs/context.html). Additional `ThemeProvider` components can be added deeper in the tree to override the original theme. The theme object will be merged into its ancestor as if by [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign). If a function is passed instead of an object it will be called with the ancestor theme and the result will be the new theme.

_Accepts:_

- **`children`: React.Node**
- **`theme`: Object|Object => Object** - An object or function that provides an object.

```jsx
import React from 'react'
import styled from '@emotion/styled'
import { ThemeProvider, withTheme } from 'emotion-theming'

// object-style theme

const theme = {
  backgroundColor: 'green',
  color: 'red'
}

// function-style theme; note that if multiple <ThemeProvider> are used,
// the parent theme will be passed as a function argument

const adjustedTheme = ancestorTheme => ({ ...ancestorTheme, color: 'blue' })

class Container extends React.Component {
  render() {
    return (
      <ThemeProvider theme={theme}>
        <ThemeProvider theme={adjustedTheme}>
          <Text>Boom shaka laka!</Text>
        </ThemeProvider>
      </ThemeProvider>
    )
  }
}
```

> Note:
>
> Make sure to hoist your theme out of render otherwise you may have performance problems.

### withTheme(component: React.ComponentType): React.ComponentType

A higher-order component that provides the current theme as a prop to the wrapped child and listens for changes. If the theme is updated, the child component will be re-rendered accordingly.

```jsx
import PropTypes from 'prop-types'
import React from 'react'
import { withTheme } from 'emotion-theming'

class TellMeTheColor extends React.Component {
  render() {
    return <div>The color is {this.props.theme.color}.</div>
  }
}

TellMeTheColor.propTypes = {
  theme: PropTypes.shape({
    color: PropTypes.string
  })
}

const TellMeTheColorWithTheme = withTheme(TellMeTheColor)
```

## Credits

Thanks goes to the [styled-components team](https://github.com/styled-components/styled-components) and [their contributors](https://github.com/styled-components/styled-components/graphs/contributors) who designed this API.

## License

MIT 2017-present
