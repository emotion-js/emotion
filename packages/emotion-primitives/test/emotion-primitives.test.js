import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import reactPrimitives from 'react-primitives'
import { ThemeProvider } from 'emotion-theming'

import emotion from 'emotion-primitives'

Enzyme.configure({ adapter: new Adapter() })

const StyleSheet = reactPrimitives.StyleSheet

jest.mock('react-primitives', () => {
  let realPrimitives = require.requireActual('react-primitives')

  return {
    ...realPrimitives,
    // mock the components to strings so that we can see the actual styles rather
    // than a bunch of class names that don't mean anything
    View: 'View',
    Text: 'Text',
    Image: 'Image'
  }
})

const theme = { backgroundColor: 'magenta' }

describe('Emotion primitives', () => {
  test('should not throw an error when used valid primitive', () => {
    expect(() => emotion.Text({})).not.toThrow()
  })

  test('should throw an error when used invalid primitive', () => {
    expect(() => emotion.TEXT({})).toThrow()
  })

  test('should render the primitive when styles applied using object style notation', () => {
    const Text = emotion.Text`color: red; font-size: 20px; background-color: ${props =>
      props.back};`
    const tree = renderer
      .create(
        <Text style={{ fontSize: 40 }} back="red">
          Emotion Primitives
        </Text>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should work with emotion-theming', () => {
    const Text = emotion.Text`color: ${props => props.theme.backgroundColor}`

    const tree = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Text>Hello World</Text>
        </ThemeProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('should render the primitive on changing the props', () => {
    const Text = emotion.Text({ padding: '20px' }, props => ({
      color: props.decor
    }))
    const tree = renderer
      .create(<Text decor="hotpink">Emotion Primitives</Text>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render primitive with style prop', () => {
    const Title = emotion.Text`color: hotpink;`
    const tree = renderer
      .create(<Title style={{ padding: 10 }}>Emotion primitives</Title>)
      .toJSON()
    expect(tree).toMatchSnapshot()
    const wrapper = Enzyme.shallow(
      <Title style={{ padding: 10 }}>Emotion primitives</Title>
    )
    expect(wrapper.find('Text').prop('style')).toEqual({
      color: 'hotpink',
      padding: 10
    })
  })

  it('should work with StyleSheet.create API', () => {
    const styles = StyleSheet.create({ foo: { color: 'red' } })
    const Text = emotion.Text`font-size: 10px;`
    const wrapper = Enzyme.shallow(
      <Text style={styles.foo}>Emotion Primitives</Text>
    )
    expect(wrapper.find('Text').prop('style')).toEqual({
      color: 'red',
      fontSize: 10
    })
  })

  test('primitive should work with `withComponent`', () => {
    const Text = emotion.Text`color: ${props => props.decor};`
    const Name = Text.withComponent(reactPrimitives.Text)
    const tree = renderer.create(<Name decor="hotpink">Mike</Name>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should style any other component', () => {
    const Text = emotion.Text`color: hotpink;`
    const Title = () => <Text>Hello World</Text>
    const StyledTitle = emotion(Title)`font-size: 20px; font-style: ${props =>
      props.sty}`
    const tree = renderer.create(<StyledTitle sty="italic" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  // it('innerRef', () => {
  //   const innerRef = jest.fn()
  //   const Text = emotion.Text`color: red;`
  //   const tree = renderer.create(<Text innerRef={innerRef}>Emotion primitives</Text>).toJSON()
  //   expect(tree).toMatchSnapshot()
  //   expect(innerRef).toBeCalled()

  //   const ref = React.createRef()
  //   const wrapper = Enzyme.mount(<Text innerRef={ref}></Text>)
  //   const text = wrapper.find('Text').first()
  //   expect(ref.current).toBe(text.instance())
  // })

  it('should pass props in withComponent', () => {
    const ViewOne = emotion.View`
      background-color: ${props => props.color};
    `
    const treeOne = renderer.create(<ViewOne color="green" />)
    const ViewTwo = ViewOne.withComponent(reactPrimitives.Text)
    const treeTwo = renderer.create(<ViewTwo color="hotpink" />)

    expect(treeOne).toMatchSnapshot()
    expect(treeTwo).toMatchSnapshot()
  })

  it('should render <Image />', () => {
    const Image = emotion.Image`
      border: 2px solid hotpink;
    `
    const tree = renderer
      .create(
        <Image
          source={{
            uri:
              'https://camo.githubusercontent.com/209bdea972b9b6ef90220c59ecbe66d35ffefa8a/68747470733a2f2f63646e2e7261776769742e636f6d2f746b6834342f656d6f74696f6e2f6d61737465722f656d6f74696f6e2e706e67',
            height: 150,
            width: 150
          }}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
