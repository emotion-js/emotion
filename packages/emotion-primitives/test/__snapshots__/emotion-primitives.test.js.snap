// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emotion primitives primitive should work with \`withComponent\` 1`] = `
<Text
  style={
    Array [
      80,
    ]
  }
>
  Mike
</Text>
`;

exports[`Emotion primitives should pass props in withComponent 1`] = `
<View
  style={
    Array [
      84,
    ]
  }
/>
`;

exports[`Emotion primitives should pass props in withComponent 2`] = `
<Text
  style={
    Array [
      84,
      Object {
        "color": "hotpink",
      },
    ]
  }
/>
`;

exports[`Emotion primitives should render <Image /> 1`] = `
<Image
  source={
    Object {
      "height": 150,
      "uri": "https://camo.githubusercontent.com/209bdea972b9b6ef90220c59ecbe66d35ffefa8a/68747470733a2f2f63646e2e7261776769742e636f6d2f746b6834342f656d6f74696f6e2f6d61737465722f656d6f74696f6e2e706e67",
      "width": 150,
    }
  }
  style={
    Array [
      85,
    ]
  }
/>
`;

exports[`Emotion primitives should render primitive with style prop 1`] = `
<Text
  style={
    Array [
      76,
      Object {
        "padding": 10,
      },
    ]
  }
>
  Emotion primitives
</Text>
`;

exports[`Emotion primitives should render the primitive on changing the props 1`] = `
<Text
  style={
    Array [
      75,
      Object {
        "color": "hotpink",
      },
    ]
  }
>
  Emotion Primitives
</Text>
`;

exports[`Emotion primitives should render the primitive when styles applied using object style notation 1`] = `
<Text
  style={
    Array [
      73,
      Object {
        "fontSize": 40,
      },
    ]
  }
>
  Emotion Primitives
</Text>
`;

exports[`Emotion primitives should style any other component 1`] = `
<Text
  style={
    Array [
      82,
    ]
  }
>
  Hello World
</Text>
`;

exports[`Emotion primitives should work with emotion-theming 1`] = `
<Text
  style={
    Array [
      74,
    ]
  }
>
  Hello World
</Text>
`;
